storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+	if (dest == FIL_NULL)
+	  goto func_exit;
+	m_extent_map[free_limit]= dest;
+	break;
+      }
+prev_extent:
+      free_limit-= FSP_EXTENT_SIZE;
+    }
+func_exit:
+    if (m_extent_map.size() == 0)
+      return false;
+
+    ib::info() << "System tablespace defragmentation process starts";
+    for (const auto &kv : m_extent_map)
+      ib::info() << "Moving the data from old extent " << kv.first
+                 << " to " << kv.second;
+    return true;
+  }
+
+  /** Defragment the system tables */
+  dberr_t defragment_system_tables()
+  {
+    dberr_t err= defragment_table(dict_sys.sys_tables);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_columns);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_indexes);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_fields);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign_cols);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_virtual);
+    return err;
+  }
+
+  /** @return extent which replaces the later extent
+  or same extent if there is no replacement exist */
+  uint32_t get_new_extent(uint32_t old_extent)
+  {
+    auto it= m_extent_map.find(old_extent);
+    if (it != m_extent_map.end())
+      return it->second;
+    return old_extent;
+  }
+
+  /** @return state for the given extent */
+  uint32_t get_state(uint32_t extent)
+  {
+    return m_extent_info[extent];
+  }
+
+  /** If the extent descriptor state is XDES_FREE then
+  allocate the new extent for the given segment.
+  If the extent descriptor state is XDES_FSEG then
+  try to find the free page from the segment.
+  @param seg_header segment header where new extent will be added
+  @param iblock     inode block
+  @param inode      index node information
+  @param extent     create a new extent
+  @param mtr        mini-transaction
+  @param err        error code
+  @return block or nullptr in case of failure */
+  buf_block_t *alloc_page_from_new_seg(
+    fseg_header_t *seg_header, buf_block_t *iblock,
+    fseg_inode_t *inode, uint32_t extent, mtr_t *mtr,
+    dberr_t *err)
+  {
+    fil_space_t *space= fil_system.sys_space;
+    buf_block_t *fsp_header= fsp_get_header(space, mtr, err);
+    if (*err)
+      return nullptr;
+    uint64_t seg_id= mach_read_from_8(inode + FSEG_ID);
+    buf_block_t *xdes;
+    xdes_t *descr=
+      xdes_get_descriptor_with_space_hdr(fsp_header, space, extent, mtr,
+                                         err, &xdes);
+    uint32_t allocated_page= FIL_NULL;
+
+    if (!descr)
+      return nullptr;
+
+    ut_a(xdes_get_state(descr) == XDES_FSEG
+         || xdes_get_state(descr) == XDES_FREE);
+
+    /* If the descriptor already belongs to the segment
+    then allocate the free page */
+    if (xdes_get_state(descr) == XDES_FSEG
+        && mach_read_from_8(descr + XDES_ID) == seg_id)
```

Comment:
The line split inside `ut_a()` is misplaced, and we are invoking the same function 3 times. Shouldn’t we avoid a crash on corrupted data?